include/dem/insertion.h

```diff
@@ -54,30 +55,72 @@ template <int dim>
+   * @param inserted_this_step Number of particles that is going to be inserted
+   * at each insetion step. This value can change in the last insertion step to
+   * reach the desired number of particles
+   * @param nx Number of insertion points in the x direction
+   * @param ny Number of insertion points in the y direction
+   * @param nz Number of insertion points in the z direction
+   * @param remained_particles Remained number of particles that should be
+   * inserted in the following steps
    */
   virtual void
-  insert(Particles::ParticleHandler<dim> &particle_handler,
-         const Triangulation<dim> &       triangulation,
-         Particles::PropertyPool &        property_pool,
-         const DEMSolverParameters<dim> & dem_parameters) = 0;
+  insert(Particles::ParticleHandler<dim> &                particle_handler,
+         const parallel::distributed::Triangulation<dim> &triangulation,
+         const DEMSolverParameters<dim> &                 dem_parameters,
+         unsigned int &                                   inserted_this_step,
+         const unsigned int &                             nx,
+         const unsigned int &                             ny,
+         const unsigned int &                             nz,
+         unsigned int &remained_particles) = 0;
+
+protected:
+  /**
+   * Carries out assigning the properties of inserted particles.
+   *
+   * @param dem_parameters DEM parameters declared in the .prm file
+   * @param inserted_this_step Number of particles that is going to be inserted
+   * at each insetion step. This value can change in the last insertion step to
+   * reach the desired number of particles
+   * @param inserted_so_far Number of particles are already inserted
+   */
+  std::vector<std::vector<double>>
+  assign_particle_propertis(const DEMSolverParameters<dim> &dem_parameters,
+                            const unsigned int &inserted_at_this_step,
+                            const unsigned int &inserted_so_far);
+
+  /**
+   * Carries out finding the insertion points of inserted particles.
+   *
+   * @param dem_parameters DEM parameters declared in the .prm file
+   * @param inserted_this_step Number of particles that is going to be inserted
+   * at each insetion step. This value can change in the last insertion step to
+   * reach the desired number of particles
+   * @param nx Number of insertion points in the x direction
+   * @param ny Number of insertion points in the y direction
+   * @param nz Number of insertion points in the z direction
+   */
+  virtual std::vector<Point<dim>>
+  assign_insertion_points(const DEMSolverParameters<dim> &dem_parameters,
+                          const unsigned int &            inserted_at_this_step,
+                          const unsigned int &            nx,
+                          const unsigned int &            ny,
+                          const unsigned int &            nz) = 0;
```

Comment:
This should be passed to the constructor of the derived classes and does not belong in the mother class.